Optimized set utilities for Matrix

albestro opened the pull request:
Close #130 
This implements optimized functions for `Matrix`, but not for `Tile`.

Add methods with relative tests (in `dlaf::matrix::util`):
- `set`
- `set_random`
- `set_random_positive_definite`

Modify `checkNear` behavior, instead of testing both absolute and relative difference, now it is possible to check even just one (by setting the other one to 0).

Minor fix on documentation.

rasolca commented on test/include/dlaf_test/util_matrix.h:
```diff
@@ -204,13 +204,11 @@ void checkPtr(PointerGetter exp_ptr, Matrix<T, Device::CPU>& mat, const char* fi
 /// @pre expected return type should be T.
-/// @pre rel_err > 0.
-/// @pre abs_err > 0.
+/// @pre rel_err > 0 || abs_err > 0
```
both has to be >= 0

albestro commented:
> Note: as it is now the diagonal are set incorrectly for the Cholesky miniapp.

@rasolca 
Right! As discussed offline, I'm working on a `set_positive_definite_hermitian` function (and its test) that, at least for the moment, it is what we need for going on with the Cholesky miniapp work.

albestro commented:
> Very handy!
> 
> Note: the header file (util_matrix.h in include/dlaf) is not in the same subdirectory of the test source code (test_util_matrix.cpp in test/unit/matrix).

@Sely85 @rasolca 
I agree with you. I think that we may decide to move everything strictly related to the matrix inside the matrix folder:
- Matrix
- Tile
- Distribution
- Matrix utils

And this should be true also for tests.

What do you think?

Sely85 commented:
> > Very handy!
> > Note: the header file (util_matrix.h in include/dlaf) is not in the same subdirectory of the test source code (test_util_matrix.cpp in test/unit/matrix).
> 
> @Sely85 @rasolca
> I agree with you. I think that we may decide to move everything strictly related to the matrix inside the matrix folder:
> 
> * Matrix
> * Tile
> * Distribution
> * Matrix utils
> 
> And this should be true also for tests.
> 
> What do you think?

I agree with you. What about `lapack_tile` and `blas_tile`? Inside the matrix folder as well since are "tile-related", right?

rasolca commented:
I agree, but since I'm working on matrix views, I propose to postpone the move of the matrix and tile headers.

albestro commented:
I've rebased and fixed conflict with master.

Trying to resume what this PR does:
- Parallelized util functions for matrixes: `set`, `set_random`, `set_random_hermitian_positive_definite` with related tests
- Introduce `Position` for generic `Index2D` with test (**WARNING**: as it is now, it permits position comparison of indexes with different types)
- Introduce `computeLinearIndex` for `Index2D`
- `CommunicatorGrid` now has a "full" communicator, with all ranks inside the grid
- Minor fix: TypeInfo now works also with `const` types
- Helper function for complex `dlaf::conj`: differently from `std::conj` that always returns an `std::complex`, this uses as return type the same as the input parameter
- Minor fix: output messages on errors for check_near for both matrix and tile

rasolca commented on include/dlaf/util_matrix.h:
```diff
@@ -146,5 +154,207 @@ void assertMultipliableMatrices(const MatrixConst& mat_a, const Matrix& mat_b, b
+          }),
+          matrix(tile_wrt_local));
+    }
+  }
```
Why you check again the position in the random getter?
(See internal comments)

Moreover this approach may lead to think that value_at({i, j}) gives always the same value for a given (i, j)

rasolca commented on include/dlaf/util_matrix.h:
```diff
@@ -146,5 +154,207 @@ void assertMultipliableMatrices(const MatrixConst& mat_a, const Matrix& mat_b, b
+                  if (i != j)
+                    tile(TileElementIndex{j, i}) = dlaf::conj(value);
+                }
+              }
```
What about the following?
```
              for (SizeType j = 0; j < tile.size().cols(); ++j) {
                for (SizeType i = 0; i < j; ++i) {
                  auto value = random();
                  tile(TileElementIndex{i, j}) = value;
                  tile(TileElementIndex{j, i}) = dlaf::conj(value);
                }
                tile({j, j}) = real(random()) + offset;
              }
```

rasolca commented on include/dlaf/util_matrix.h:
```diff
@@ -146,5 +154,207 @@ void assertMultipliableMatrices(const MatrixConst& mat_a, const Matrix& mat_b, b
+                  else
+                    tile(TileElementIndex{j, i}) = value;
+                }
+              }
```
What about the following?
```
              for (SizeType j = 0; j < tile.size().cols(); ++j) {
                for (SizeType i = 0; i < tile.size().rows(); ++i) {
                  if (tile_position == dlaf::common::Position::LOWER)
                    tile(TileElementIndex{i, j}) = random();
                  else
                    tile(TileElementIndex{j, i}) = conj(random());
                }
              }
```

rasolca commented on include/dlaf/common/index2d.h:
```diff
@@ -171,6 +171,25 @@ template <class Index2DType, typename LinearIndexT>
+  else if (index.row() < index.col())
+    return Position::UPPER;
+  return Position::DIAGONAL;
+}
```
This may may be misleading:
A tile that contains only elements of the lower triangular part but has an index (i, j) where j>i is considered UPPER.

rasolca replied:
Moreover a simple check `if (index.row() <= index.col())`
becomes `if (position(index) >= Diagonal)` which
requires 2 condition evaluations in case of false and 3 in case of true.

rasolca commented on include/dlaf/util_matrix.h:
```diff
@@ -12,6 +12,12 @@
 #include <string>
+#include <cmath>
+
+#undef M_PI
```
First you un-define M_PI and than you check if it is defined?

rasolca commented on include/dlaf/util_matrix.h:
```diff
@@ -12,6 +12,12 @@
+#include <cmath>
+
+#undef M_PI
+#if !defined M_PI
```
Personally I prefer `#ifndef`.

rasolca commented on test/unit/communication/test_communicator_grid.cpp:
```diff
@@ -187,7 +187,7 @@ TEST_P(CommunicatorGridTest, ConstructorIncomplete) {
     EXPECT_EQ(MPI_COMM_NULL, incomplete_grid.colCommunicator());
 
-    EXPECT_EQ(-1, incomplete_grid.rank_all(coords));
+    EXPECT_EQ(-1, incomplete_grid.rankFullCommunicator(coords));
```
Shouldn't `rankFullCommunicator` be tested with different values of coords? (applies also to other places in the code)

rasolca commented on include/dlaf/util_matrix.h:
```diff
@@ -146,5 +159,179 @@ void assertMultipliableMatrices(const MatrixConst& mat_a, const Matrix& mat_b, b
+  }
+}
+
+/// Set the matrix with random values in the range [-1, 1]
```
What about the complex case?
